### DISCUSSION TITLE
Custom login/register page for keycloak

### DISCUSSION THREAD

**Discussion by aditodkar** [keycloakify/keycloakify#52] [Q&A]
I am facing issue adding react login page in keycloak can someone please help and let me know what I am missing or is there something which I am doing wrong?? 
Code details: https://keycloak.discourse.group/t/custom-login-ans-signup-page-instead-of-custom-theme/1792

**Comment by aditodkar** [comment]
> I am facing issue adding react login page in keycloak can someone please help and let me know what I am missing or is there something which I am doing wrong?? 
> Code details: https://keycloak.discourse.group/t/custom-login-ans-signup-page-instead-of-custom-theme/1792

Getting this message as well

🔏 Building the keycloak theme...⌚️
'rm' is not recognized as an internal or external command,
operable program or batch file.
D:\Keycloak\myapp\node_modules\keycloakify\bin\download-builtin-keycloak-theme.js:37
        finally { if (e_1) throw e_1.error; }
                           ^

Error: Command failed: rm -rf D:\Keycloak\myapp\build_keycloak\src\main\resources\theme\myapp\tmp_xxKdOxnEdx
'rm' is not recognized as an internal or external command,
operable program or batch file.

**Reply by garronej** [reply]
Hi,
You can't build on windows.
You need WSL

**Reply by aditodkar** [reply]
If I try to run 
then I get following message: `./build_keycloak/start_keycloak_testing_container.sh`
```
Error: No such container: keycloak-testing-container
[+] Building 11.0s (2/3)                                                                                                                                                  
[+] Building 11.0s (3/3) FINISHED                                                                                                                                         
 => [internal] load build definition from Dockerfile                                                                                                                 0.0s
 => => transferring dockerfile: 38B                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                    0.0s
 => => transferring context: 2B                                                                                                                                      0.0s
 => ERROR [internal] load metadata for docker.io/jboss/keycloak:15.0.2                                                                                              10.9s
------
 > [internal] load metadata for docker.io/jboss/keycloak:15.0.2:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: rpc error: code = Unknown desc = error getting credentials - err: exit status 1, out: ``
Unable to find image 'myapp/keycloak-hot-reload:latest' locally
<3>init: (8357) ERROR: UtilConnectToInteropServer:307: connect failed 2
docker: Error response from daemon: pull access denied for myapp/keycloak-hot-reload, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'
```
So this issue is specific to vscode in windows WSL so I ran  `./build_keycloak/start_keycloak_testing_container.sh` command in terminal instead of vscode and it worked fine. 

Now one problem is that when this command is executed successfully if I navigate to localhost:8080 then after login it shows custom theme  (i.e myapp create-react-app name) added in theme dropdown in admin console but it is master realm instead of that I want to have that custom theme on another realm how can I do that? (In short theme is only reflecting on master realm instead of realm which I have created)

**Reply by garronej** [reply]
I don't understand...

https://user-images.githubusercontent.com/6702424/142637593-359f873e-d944-47f2-8a79-6d0bdbf06fa3.mov

**Reply by garronej** [reply]
You can even select the login theme for a specific client...

https://user-images.githubusercontent.com/6702424/142637891-6c076601-a247-4c20-a934-78627ffe80fc.mov

**Reply by aditodkar** [reply]
Let me explain or give some more info. I have docker keycloak running with postgreSQL  [docker-compose file](https://github.com/keycloak/keycloak-containers/blob/main/docker-compose-examples/keycloak-postgres.yml) now keycloak is running in background on port 8080 and I have created client and users already. Now I have created a separate create-react-app with all the required changes in package.json. Now I run this command `npm run keycloak` which generates build_keycloak folder (see screenshot below). In console is gives some info like you can run the test container to see custom login page by running `./build_keycloak/start_keycloak_testing_container.sh`

console message for more info:
```
To test your theme locally, with hot reloading, you can spin up a Keycloak container image with the theme loaded by running:

👉 $ ./build_keycloak/start_keycloak_testing_container.sh 👈

To enable the theme within keycloak log into the admin console ( 👉 http://localhost:8080 username: XXXX, password: XXXXX👈), create a realm (called "myrealm" for example),
go to your realm settings, click on the theme tab then select myapp.
More details: https://www.keycloak.org/getting-started/getting-started-docker

Once your container is up and configured 👉 http://localhost:8080/auth/realms/myrealm/account 
```

I am not sure but do I need to copy these theme files to keycloak docker container and only then it will reflect in admin console under theme dropdown? Because right now if I go to client ->  theme dropdown I don't see myapp in theme dropdown. Please suggest if I am missing something. Thanks in advance :) 

![Screenshot_2](https://user-images.githubusercontent.com/16613832/142640646-5ff074e1-368d-405e-a253-630eae7e6e0b.png)

**Reply by garronej** [reply]
Ahhhhh ok!
`start_keycloak_testing_container.sh` is just for testing. 
The actual theme is the `jar` file. This is what you are supposed to load in your Keycloak instance.  

![image](https://user-images.githubusercontent.com/6702424/142642744-9aafa2b9-a8ef-40d0-b230-353852caacb4.png)

You have to put it in [`/opt/jboss/keycloak/standalone/deployments/`](https://github.com/InseeFrLab/keycloakify/blob/f2eadf54414cab2e6833d7add6d4b89a451f1ac4/src/bin/build-keycloak-theme/build-keycloak-theme.ts#L103)

**Reply by aditodkar** [reply]
So do you mean to say once I generate this keycloakify jar file then I have to run keycloak docker container in interactive mode and then copy this jar file inside `/opt/jboss/keycloak/themes/myapp` is that right?

Something like this:
1) First run keycloak docker container in interactive mode.
`docker exec -it container_ID bin/bash`
2) `mkdir /opt/jboss/keycloak/themes/myapp`
3) then do `docker cp myapp/. container_ID:/opt/jboss/keycloak/themes/myapp`
4) and then docker commit 

Is this correct approach or is there any better way to do this? Please suggest :)

**Reply by garronej** [reply]
Sorry it's in `/opt/jboss/keycloak/standalone/deployments/` that you must put the `.jar` file indeed.  

As for how you should package it, it's up to you but it should be a line like this in your `Dockerfile`:  
`COPY  my-react-app-theme.jar /opt/jboss/keycloak/standalone/deployments/`

**Reply by aditodkar** [reply]
1) I was referring this video https://youtu.be/Xg5y57BedoE?t=295 check at 4:55 min that guy generates jar file and puts it inside providers.
2) In this blog https://austincunningham.ddns.net/2020/changekeycloakdockertheme it is copied to themes folder 
So I got confused exactly where to put

**Reply by garronej** [reply]
> In this blog https://austincunningham.ddns.net/2020/changekeycloakdockertheme it is copied to themes folder
So I got confused exactly where to put

In this article the author doesn't have a `.jar` file like you have. He is copying every files.  

Once you'll have your Dockerfile working it would be nice to paste it here for helping the community.

**Reply by aditodkar** [reply]
` --> Once you'll have your Dockerfile working it would be nice to paste it here for helping the community.`
Yes I am going to write detailed article after this :) 
I have one question lets say CRA and keycloak is running on docker. Once I create jar file inside build_keycloak folder then can I move that jar file to keycloak container at this path: `/opt/jboss/keycloak/standalone/deployments/`
In short after each code change is it possible to generate new jar file and move it to keycloak server container at this path: `/opt/jboss/keycloak/standalone/deployments/`

If it can be done then will hot reloading as well work?

docker-compose.yml: https://gist.github.com/aditodkar/ffdaf49e7f7ae6112c79ac5e237ef602#file-docker-compose-yml-L9

**Reply by garronej** [reply]
> Yes I am going to write detailed article after this :)

Looking forward to read your article!  

> I have one question lets say CRA and keycloak ...

Sure it's possible. 
At InseeFrLab we have a continuous delivery pipeline set up for our app [onyxia](https://github.com/InseeFrLab/onyxia).  

The first thing is to setup a CI workflow that will build the Docker image for the CRA app and the `.jar` automatically.
This is documented in the template [keycloakiy-demo-app](https://github.com/garronej/keycloakify-demo-app).

Once we have that, the question is: how do we automatically update our app in production.  

Well, it depends of your production setup. At InseeFrLab we use Kubernetes + ArgoCD.
We have a private Helm repo containing all the configuration of our production apps. Every time something changes on this repo ArgoCD restarts the relevant pods.  

Here is our Helm configuration for Keycloak:

`apps/keycloak/values.yaml`
```yaml
keycloak:
  # The number of replicas to create (has no effect if autoscaling enabled)
  replicas: 2

  extraInitContainers: |
    - name: realm-ext-provider
      image: curlimages/curl
      imagePullPolicy: IfNotPresent
      command:
        - sh
      args:
        - -c
        - |
          curl -L -f -S -o /extensions/onyxia-web.jar https://github.com/InseeFrLab/onyxia-web/releases/download/v0.19.9/keycloak-theme.jar
      volumeMounts:
        - name: extensions
          mountPath: /extensions
  extraVolumeMounts: |
    - name: extensions
      mountPath: /opt/jboss/keycloak/standalone/deployments
  extraVolumes: |
    - name: extensions
      emptyDir: {}
  # Additional environment variables for Keycloak
  extraEnv: |
    - name: KEYCLOAK_USER
      value: superuser
    - name: KEYCLOAK_PASSWORD
      value: xxxxxxxxxxxxxxxxxx
    - name: JAVA_OPTS
      value: >-
        -Dkeycloak.profile=preview
```

`apps/keycloak/Chart.yaml`

```yaml
apiVersion: v2
name: keycloak
version:  1.0.33
dependencies:
  - name: keycloak
    version: 15.1.0
    repository: https://codecentric.github.io/helm-charts
    enabled: true
```

You see that when the container starts we download the `.jar` of our theme  and put it in `/opt/jboss/keycloak/standalone/deployments`.

Every times we want to deploy a new version of [onyxia-web](https://github.com/InseeFrLab/onyxia-web) we update the `package.json` version and we push. It triggers our CI workflow that is very much alike the one descibed in keycloakify-demo-app but with [an extra step](https://github.com/InseeFrLab/onyxia-web/blob/5f6f83f7f94d2673733975cfe9c416a45bc1fd5a/.github/workflows/ci.yml#L169-L177) that triggers the CI of our gitops repo. 
The CI of our gitops repo checks if there is a new version of onyxia-web available. If there is, it makes a commit to itself to changes the versions number of the CRA app and of the `.jar` being downloaded and installed at the container init:  

![image](https://user-images.githubusercontent.com/6702424/142789044-341b382f-e32f-41a5-809e-18f6c4b3d341.png)
(If you want the code of the GitHub Actions that perform this automatic commit, I can give it to you.)

After that ArgoCD restarts our pods and [our app in production](https://datalab.sspcloud.fr) is updated.
